src/CMakeLists.txt

```diff
@@ -53,43 +56,41 @@ foreach (sourcefile ${share_sources})
   endif()
 endforeach()
 
-# Remove shr_cal_mod from share_sources.
-#
-# shr_cal_mod depends on ESMF (or the lightweight esmf wrf timemgr, at
-# least). Since CTSM doesn't currently use shr_cal_mod, we're avoiding
-# the extra overhead of including esmf_wrf_timemgr sources in this
-# build.
-#
-# TODO: like above, this should be moved into a general-purpose function
-# in Sourcelist_utils.  Then this block of code could be replaced with a
-# single call, like: remove_source_file(${share_sources}
-# "shr_cal_mod.F90")
-foreach (sourcefile ${share_sources})
-  string(REGEX MATCH "shr_cal_mod.F90" match_found ${sourcefile})
-  if(match_found)
-    list(REMOVE_ITEM share_sources ${sourcefile})
-  endif()
-endforeach()
+if (DEFINED PIO)
+  set(PIO_PATH ${PIO})
+else()
+  set(PIO_PATH $ENV{PIO})
+endif()
```

Comment:
Here and some related stuff below: Do we truly need PIO? I thought we had been taking pains to avoid dependence on PIO in the unit test build.

At a glance it looks like this relies on having PIO pre-installed on the system, as opposed to building it from source which I believe is still an option in the CESM build.

Reply:
The need for PIO and NETCDF is because ESMF requires it. It wouldn't link without having PIO and NetCDF libraries available. It's maybe possible to use a different ESMF version that doesn't have I/O -- but that would mean having a different ESMF than the main CESM build uses. And I don't think we want that.

You are right that we don't actually want I/O in our PF unit tests (and I confirmed this with the group last week). But, not having to do stubs for modules that have I/O might actually make it easier to add new PF unit tests.

So I think this is OK, and maybe a benefit. Does that sound acceptable to you?

Reply:
Resolving this.

Reply:
Thanks for the explanation about the need for PIO. This makes sense now.

My main concern is that this will make it somewhat harder for people to build the unit tests on their own systems. Now they will need to maintain PIO as a separately-built library rather than letting it be built as part of the CESM build process. But maybe that's the direction things are (or should be) moving anyway, so I guess this is okay. My understanding is that currently you have a choice: you can either (1) Build PIO on your system, then build ESMF in a way that uses this external PIO, then build CESM in a way that links against the same external PIO; or (2) Do *not* bother building PIO on your system, then build ESMF in a way that uses its own internal version of PIO (which is something close to the latest PIO release), then build CESM in a way that uses its own internal version of PIO. (2) is simpler and is what I've done up until now on my local system, but can theoretically cause problems if the PIO versions in ESMF and CESM are incompatible in some way.

ESMF provides a way to query the relevant build options, including libraries that should be included in the link like pio, via ESMFMKFILE, but I'm not sure if there's a way to easily ingest this into a CMake script.

That's all to say: I don't love that it will now be a little more complex for people to maintain their own systems for the sake of unit testing, and especially that supported ways of building the full CESM aren't supported for unit testing (which, if anything, should be simpler rather than more complex). But the reasons for this make sense, and arguably we shouldn't be supporting the use of internal PIO in the ESMF and CESM builds anyway, so I'm okay with this.

Reply:
That makes sense @billsacks. I did tell @adrifoster that she'll need to build ESMF, PIO, and NetCDF on her system. She already has ESMF and NetCDF, but will need to make sure PIO is also in the mix now. Although now that you mention it -- she must have PIO somewhere -- so maybe was using PIO that's part of ESMF? In her case she also wanted to get regular CESM cases to work, and there you do need all those things anyway. So it's only a problem if you want to do unit-testing easily -- but do NOT need to run CESM cases. But, you are right it's good to reduce complexity when it isn't necessary.

So you are right it isn't great, but seems like the reasonable way forward right now. Thanks for your comments and review.

Reply:
@ekluzek and @adrifoster - yes, I think @adrifoster probably had things set up similarly to me, which hasn't required any separate / pre-built PIO, instead using the default internal PIO both for ESMF and for CESM.

@ekluzek , you say, "In her case she also wanted to get regular CESM cases to work, and there you do need all those things anyway." -- but I think that's not actually true, because I think it's still possible to use the internal PIO for regular CESM cases, whereas starting with this PR that won't be possible with the unit testing.

Again, I think that's probably okay, since it seems like the way we're moving anyway, but it does add one more bit of complexity for setting up and maintaining a system for CTSM development.

One possible way around this would be to make all of this PIO addition conditional: If PIO isn't defined in your machine setup, I think it's safe to assume that you're using internal PIO in both ESMF and CESM on your system, and you should be able to bypass the PIO stuff you added here - at least until we actually need PIO in the CTSM unit tests for their own sake, beyond just ESMF. I'm not sure it's worth doing that, but I'm raising it as an option, e.g., if @adrifoster finds it's challenging to set up or maintain the PIO build.

Reply:
I wanted to add a bit more information about linking ESMF related to the earlier discussion about the current hard-coding of linking with PIO: Based on some discussion with Dan Rosen on the ESMF team, it sounds like it should be possible to get ESMF's required link libraries if you use ESMF's provided FindESMF.cmake: that provides INTERFACE_LINK_LIBRARIES, which I think are the extra link flags (paths, libs and options) needed when linking ESMF. (And if that doesn't provide what's needed, then everything in the esmf.mk file is available in CMake after calling FindESMF.cmake, but currently that requires some parsing / manipulation of the settings there.) Generally you would use a CMake Find module like this to actually find the package; in this case we already know where to find ESMF, so it's a little funny to use the Find module, but you can still use it to get various variables like this.

I think that, in the interest of moving things along, it's fine to stick with what you have for now, but this is a note as something that could clean this up and make it more machine-portable long-term.